# The overall objective of this course is to enable you to perform a differential gene expression analysis from start to finish. <br>

To achieve this, we will develop the following skills: <br>
* become familiar with the command-line (Linux) environment,
  including navigating the directory structure, making, viewing and editing files,
  and learning other basic SHELL/BASH commands.
* perform more sophisticated command-line operations such as using regular expressions (regex),
  piping, and for-loops to efficiently manipulate and reformat/prepare data for RNASeq analysis.
* employ commonly-used software packages to process fastq files (including adapter removal and quality filtering),
  and generate summary statistics.
* quantify read counts via transcriptomic mapping approaches.
* analyze read quantitation data to detect differential gene expression and identify significantly enriched pathways.
* visualize DEG and pathway enrichment results with various R packages.
